### ISSUE TITLE
add_index with an expression ignores opclass specification

### ISSUE THREAD

**Issue by DaanVanVugt** [rails/rails#53227]
### Steps to reproduce
1. Generate a migration like:
```ruby
class AddFullNameIndexToUsers < ActiveRecord::Migration[7.1]
  def change
    expr = "(first_name || ' ' || last_name)"
    add_index :users,
              expr,
              name: 'index_users_on_full_name_trigram',
              using: :gin,
              opclass: { expr => :gin_trgm_ops }
  end
end
```

### Expected behavior
An index is created and stored in schema.rb

### Actual behavior
> PG::UndefinedObject: ERROR:  data type text has no default operator class for access method "gin" (PG::UndefinedObject)
HINT:  You must specify an operator class for the index or define a default operator class for the data type.

The SQL query that is executed is:

>  CREATE INDEX "index_users_on_full_name_trigram" ON "users" USING gin ((first_name || ' ' || last_name))

### System configuration
**Rails version**:
7.1.3.4
**Ruby version**:
3.3.0

**Comment by fatkodima** [comment]
You need to also create the required extensions:
```ruby
    enable_extension "pg_trgm"
    enable_extension "btree_gin"
```

**Comment by DaanVanVugt** [comment]
I don't think this is the solution. pg_trgm is enabled, and I can create a trigram gin index for non-computed columns. It seems to me that the problem is that `quoted_columns_for_index` is not called because the computed column name is a string, at https://github.com/rails/rails/blob/main/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L134

**Comment by fatkodima** [comment]
Can you create a reproduction test script based on https://github.com/rails/rails/blob/main/guides/bug_report_templates/active_record_migrations.rb template?

**Comment by DaanVanVugt** [comment]
I have succeeded in creating the index with the following (a little ugly) migration:

```ruby
class AddFullNameIndexToUsers < ActiveRecord::Migration[7.1]
  def change
    expr = "(first_name || ' ' || last_name) gin_trgm_ops"
    add_index :users,
              expr,
              name: 'index_users_on_full_name_trigram',
              using: :gin
  end
end
```

which produces the following SQL:
```sql
CREATE INDEX "index_users_on_full_name_trigram" ON "users" USING gin ((first_name || ' ' || last_name) gin_trgm_ops)
```

and results in this line in `schema.rb`, which works correctly with a schema load.

```ruby
t.index "((((first_name)::text || ' '::text) || (last_name)::text)) gin_trgm_ops", name: "index_users_on_full_name_trigram", using: :gin
```

it is not very pretty, and perhaps it would be good to have a line in the documentation about this being possible.

**Comment by fatkodima** [comment]
I was able to run the migration successfully with
```ruby
class AddFullNameIndexToUsers < ActiveRecord::Migration[7.1]
  def change
    enable_extension "pg_trgm"
    enable_extension "btree_gin"

    expr = "(first_name || ' ' || last_name)"
    add_index :users,
              expr,
              name: 'index_users_on_full_name_trigram',
              using: :gin,
              opclass: { expr => :gin_trgm_ops }
  end
end
```

Thats why I closed this issue.

**Comment by DaanVanVugt** [comment]
I think perhaps btree_gin sets the default opclass for text, which would also work. Probably this last migration executes slightly different SQL.
